Corrections and clarifications for FILE_RENAME_INFO

idigdoug opened the pull request:
Remove duplicated ReplaceIfExists field.

Since ReplaceIfExists and Flags are in a union, we need to explain when each one is used. ReplaceIfExists is used for FileRenameInfo, and Flags is used for FileRenameInfoEx.

Ideally the accepted values for Flags should be documented (they're listed in https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/ntifs/ns-ntifs-_file_rename_information), but I'm not sure which flags should be shown here or the version of windows when each flag becomes valid, so I'm not going to add that.

Clarify the usage of RootDirectory. Most users should set it to NULL. Wording is taken from https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/ntifs/ns-ntifs-_file_rename_information.

FileNameLength needs to be clearly documented as the FileName size in bytes without the NUL termination. (Technical note: at present this value is only used in some code paths. In other code paths, FileName needs to be nul-terminated.)

Clarify that FileName MUST be NUL-terminated. Some code paths use the NUL-termination and other code paths use FileNameLength, so both must be set correctly.

idigdoug committed:
Corrections and clarifications for FILE_RENAME_INFO

Remove duplicated ReplaceIfExists field.

Since ReplaceIfExists and Flags are in a union, we need to explain when each one is used. ReplaceIfExists is used for FileRenameInfo, and Flags is used for FileRenameInfoEx.

Ideally the accepted values for Flags should be documented (they're listed in https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/ntifs/ns-ntifs-_file_rename_information), but I'm not sure which flags should be shown here or the version of windows when each flag becomes valid, so I'm not going to add that.

Clarify the usage of RootDirectory. Most users should set it to NULL. Wording is taken from https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/ntifs/ns-ntifs-_file_rename_information.

FileNameLength needs to be clearly documented as the FileName size in bytes without the NUL termination. (Technical note: at present this value is only used in some code paths. In other code paths, FileName needs to be nul-terminated.)

Clarify that FileName MUST be NUL-terminated. Some code paths use the NUL-termination and other code paths use FileNameLength, so both must be set correctly.

PRMerger4 commented:
@idigdoug : Thanks for your contribution! The author(s) have been notified to review your proposed change.

idigdoug committed:
FileName is relative to current directory, not handle.

Additional investigation indicates that the FileName is relative to the current directory, not the source file handle. Clarify that in the documentation.

PRMerger7 commented:
@idigdoug : Thanks for your contribution! The author(s) have been notified to review your proposed change.
